dask_cuda/explicit_comms/dataframe/shuffle.py

```diff
@@ -258,27 +259,37 @@ def shuffle(
         "explicit-comms-shuffle-group-"
         f"{tokenize(df, column_names, npartitions, ignore_index)}"
     )
-    df = df.persist()  # Making sure optimizations are apply on the existing graph
-    dsk = dict(df.__dask_graph__())
-    output_keys = []
-    for input_key in df.__dask_keys__():
-        output_key = (name, input_key[1])
-        dsk[output_key] = (
+    # Convert to bag and use map_partitions so we can enable
+    # blockwise task fusion (if df is not alrady persisted)
+    if hasattr(df._meta, "partition_by_hash"):
+        # Use partition_by_hash for cudf-backed data
+        # (may be faster with an rmm pool)
+        def _shuffle_group(x, **kwargs):
+            return {i: part for i, part in enumerate(x.partition_by_hash(**kwargs))}
+
+        df_groups = df.to_bag(format="frame").map_partitions(
+            _shuffle_group,
+            columns=column_names,
+            nparts=npartitions,
+            keep_index=not ignore_index,
+        )
```

Comment:
Yeah, unfortunately, `df_groups.compute()` will do something very different than `df_groups.compute(sync=False)` (and the latter *should* probably give you an error without distributed). The plain `compute` call will hit post-compute logic in `dask.bag`, which will inadvertently extract the keys from the dictionary of DataFrame pieces. The `sync=False` version should skip the post-compute logic, because the result should be a list of distinct `Futures`. Calling `*.result()` on any of these Futures should return the `dict` that we are expecting.

Reply:
Is the non-distributed (non-sync) case right? It seems like very odd behaviour.

Reply:
> It seems like very odd behaviour.

It does feel **wrong** to me as well, but it is consistent with the general behavior of Dask-Bag. That is, Dask-Bag assumes you are partitioning something like a *sequence* of records (not a single **large** dictionary).  Therefore, I suppose `df_groups` isn't really a *true* DataFrame or a Bag collection. The ony rason Bag is considered "more correct" is that DataFrame metadata validation will fail (unless you "lie" about meta), while Bag does't care about the underlying data until you call `compute`.

Reply:
OK, thanks